pkg/util/billing/billing.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return nil, err
+	}
+	key := *(*keys.Keys)[0].Value
+
+	client, err := azstorage.NewBasicClient(env.E2EStorageAccountName(), key)
+	if err != nil {
+		return nil, err
+	}
+
+	return &manager{
+		storageClient: &client,
+		subDB:         sub,
+		billingDB:     billing,
+		log:           log,
+	}, nil
+}
+
+func (m *manager) Create(ctx context.Context, doc *api.OpenShiftClusterDocument) error {
+	billingDoc, err := m.billingDB.Create(ctx, &api.BillingDocument{
+		ID:                        doc.ID,
+		Key:                       doc.Key,
+		ClusterResourceGroupIDKey: doc.ClusterResourceGroupIDKey,
+		Billing: &api.Billing{
+			TenantID: doc.OpenShiftCluster.Properties.ServicePrincipalProfile.TenantID,
+			Location: doc.OpenShiftCluster.Location,
+		},
+	})
+	if err, ok := err.(*cosmosdb.Error); ok &&
+		err.StatusCode == http.StatusConflict {
+		m.log.Print("billing record already present in DB")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if e2eErr := m.createOrUpdateE2EBlob(ctx, billingDoc); e2eErr != nil {
+		m.log.Warnf("createOrUpdateE2EBlob failed: %s", e2eErr)
+	}
+
+	return nil
+}
+
+func (m *manager) Delete(ctx context.Context, doc *api.OpenShiftClusterDocument) error {
+	m.log.Printf("updating billing record with deletion time")
+	billingDoc, err := m.billingDB.MarkForDeletion(ctx, doc.ID)
+	if cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if e2eErr := m.createOrUpdateE2EBlob(ctx, billingDoc); e2eErr != nil {
+		// We are not failing the operation if we cannot write to e2e storage account, just warning
+		m.log.Warnf("createOrUpdateE2EBlob failed: %s", e2eErr)
+	}
+
+	return nil
+}
+
+// isSubscriptionRegisteredForE2E returns true if the subscription has the
+// "Microsoft.RedHatOpenShift/SaveAROTestConfig" feature registered
+func isSubscriptionRegisteredForE2E(sub *api.SubscriptionProperties) bool {
+	if sub.TenantID == tenantIDMSFT || sub.TenantID == tenantIDAME {
+		return feature.IsRegisteredForFeature(sub, featureSaveAROTestConfig)
+	}
+	return false
+}
+
+// createOrUpdateE2Eblob create a copy of the billing document in the e2e
+// storage account. This is used later on by the billing e2e
+func (m *manager) createOrUpdateE2EBlob(ctx context.Context, doc *api.BillingDocument) error {
+	// Validate if E2E Feature is registered
+	resource, err := azure.ParseResourceID(doc.Key)
+	if err != nil {
+		return err
+	}
+
+	subDocument, err := m.subDB.Get(ctx, resource.SubscriptionID)
+	if err != nil {
+		return err
+	}
+
+	if !isSubscriptionRegisteredForE2E(subDocument.Subscription.Properties) {
+		return nil
+	}
+
+	blobclient := m.storageClient.GetBlobService()
+
+	containerName := strings.ToLower("bill-" + doc.Billing.Location + "-" + resource.ResourceGroup + "-" + resource.ResourceName)
+	if len(containerName) > 63 {
+		containerName = containerName[:63]
```

Comment:
@JackQuincy this was 62 - I'm presuming an off-by-one error

Reply:
yes it was an off by one error